### ISSUE TITLE
categorical_crossentropy's output is non-zero when y_true and y_pred are 1

### ISSUE THREAD

**Issue by drewshark** [tensorflow/tensorflow#60113]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

Yes

### Source

source

### Tensorflow Version

2.11.0

### Custom Code

Yes

### Current Behaviour?

```shell
When `y_true` and `y_pred` are `1.`, the categorical crossentropy outputted is a non-zero value. But it should be **zero** following the design of categorical crossentropy.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

y_true = tf.constant([[1.]])
y_pred = tf.constant([[1.]])

loss = tf.keras.metrics.categorical_crossentropy(y_true, y_pred)
print(loss)  # 1.1920929e-07 but should be 0
```
```


### Relevant log output

```shell
1.1920929e-07
```
</details>

**Comment by pat749** [comment]
The issue with this code is that it is using '**categorical_crossentropy**' loss function for a binary classification problem. '**categorical_crossentropy**' is meant for multi-class classification problems where each example can belong to one of several classes.
`import tensorflow as tf

y_true = tf.constant([[1.]])
y_pred = tf.constant([[1.]])

loss = tf.keras.losses.binary_crossentropy(y_true, y_pred)
print(loss)  # 0.0
`

**Comment by pat749** [comment]
### Multi class error

The issue with this code is that it is using **categorical_crossentropy** loss function for a binary classification problem.**categorical_crossentropy**is meant for multi-class classification problems where each example can belong to one of several classes.

For a binary classification problem with only two classes (0 or 1), binary_crossentropy is the appropriate loss function to use.

```
import tensorflow as tf

y_true = tf.constant([[1.]])
y_pred = tf.constant([[1.]])

loss = tf.keras.losses.binary_crossentropy(y_true, y_pred)
print(loss)  # 0.0
```

**Comment by neel-jotaniya** [comment]
@pat749 i think it because limitations of floating-point arithmetic used by computers.

Obtaining a very small non-zero loss value even when the actual and predicted values are the same is not a bug, it is a normal behavior and expected in most cases.

**Comment by pat749** [comment]
@neel-jotaniya That is correct. In many cases, such as in machine learning models, the goal is to minimize the difference between the predicted and actual values, which is usually measured using a loss function. However, it is often impossible to achieve a perfect match between the predicted and actual values, so a very small non-zero loss value is expected and acceptable. This is particularly true for regression problems where the output is a continuous variable.

**Comment by sachinprasadhs** [comment]
Closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
